source/extensions/dynamic_modules/sdk/rust/src/lib.rs

```diff
@@ -206,360 +209,385 @@ impl EnvoyHttpFilterConfig {
+      let key_size = key.len();
+
       let mut result_ptr: *const u8 = std::ptr::null();
       let mut result_size: usize = 0;
+
       unsafe {
         callback(
           self.raw_ptr,
           key_ptr as *const _ as *mut _,
           key_size,
           &mut result_ptr as *mut _ as *mut _,
           &mut result_size as *mut _ as *mut _,
-          i,
+          0, // Only the first value is needed.
         )
       };
-      // Within the range, all results are guaranteed to be non-null by Envoy.
-      results.push(unsafe { std::slice::from_raw_parts(result_ptr, result_size) });
+
+      if result_ptr.is_null() {
+        None
+      } else {
+        Some(EnvoyBuffer::new(result_ptr, result_size))
+      }
+    }
+
+    /// This implements the common logic for getting the header/trailer values.
+    ///
+    /// TODO: use smallvec or similar to avoid the heap allocations for majority of the cases.
+    fn get_header_values_impl(
+      &self,
+      key: &str,
+      callback: unsafe extern "C" fn(
+        filter_envoy_ptr: abi::envoy_dynamic_module_type_http_filter_envoy_ptr,
+        key: abi::envoy_dynamic_module_type_buffer_module_ptr,
+        key_length: usize,
+        result_buffer_ptr: *mut abi::envoy_dynamic_module_type_buffer_envoy_ptr,
+        result_buffer_length_ptr: *mut usize,
+        index: usize,
+      ) -> usize,
+    ) -> Vec<EnvoyBuffer> {
+      let key_ptr = key.as_ptr();
+      let key_size = key.len();
+      let mut result_ptr: *const u8 = std::ptr::null();
+      let mut result_size: usize = 0;
+
+      // Get the first value to get the count.
+      let counts = unsafe {
+        callback(
+          self.raw_ptr,
+          key_ptr as *const _ as *mut _,
+          key_size,
+          &mut result_ptr as *mut _ as *mut _,
+          &mut result_size as *mut _ as *mut _,
+          0,
+        )
+      };
+
+      let mut results = Vec::new();
+      if counts == 0 {
+        return results;
+      }
+
+      // At this point, we assume at least one value is present.
+      results.push(EnvoyBuffer::new(result_ptr, result_size));
+      // So, we iterate from 1 to counts - 1.
+      for i in 1 .. counts {
+        let mut result_ptr: *const u8 = std::ptr::null();
+        let mut result_size: usize = 0;
+        unsafe {
+          callback(
+            self.raw_ptr,
+            key_ptr as *const _ as *mut _,
+            key_size,
+            &mut result_ptr as *mut _ as *mut _,
+            &mut result_size as *mut _ as *mut _,
+            i,
+          )
+        };
+        // Within the range, all results are guaranteed to be non-null by Envoy.
+        results.push(EnvoyBuffer::new(result_ptr, result_size));
+      }
+      results
     }
-    results
   }
 }
 
+#[cfg(not(any(test, feature = "testing")))]
+pub type EnvoyHttpFilter = envoy_http_filter::EnvoyHttpFilter;
+#[cfg(any(test, feature = "testing"))]
+pub type EnvoyHttpFilter = envoy_http_filter::MockEnvoyHttpFilter;
```

Comment:
The fact that automock works on structs is nice in theory but it does require double compiling which IMO is kind of annoying in practice, so I never use it in my own projects. I guess if we do it this way we have to have the testing feature per my other comment and default it to off. I guess we can see how this goes and if anyone complains.

Reply:
i came to the same conclusion that mocking a struct makes this unnecessarily complex and i switched to the trait mocking in https://github.com/envoyproxy/envoy/pull/37753/commits/9207f36f0bc0f430a002883d47740d24ed124edc - even though it came with a type parameter plumbing, i think it's simpler as users can use the mock by default